### PR TITLE
plasma-infra: improve cypress test

### PR DESCRIPTION
**Pull request by Yakutoc**
Решили ускорить через параллельный запуск.

 Screenshot 2023-06-14 at 13 23 38

#### Плюсы: 

Стало быстрее. В среднем теперь прогон **всех** cypress тестов занимает около **20-23 мин** против 39-47 мин. 

Screenshot 2023-06-14 at 13 28 10

В **худшем** случае 27-30 мин.
       
Screenshot 2023-06-14 at 13 28 43

#### Минусы: 

Если раньше данный workflow был монолитом и выполнялся в рамках одной job, теперь их 8.

То есть можем упереться в ресурсы всего CI.

:baby_chick: Download canary assets:

[color_asdk_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_asdk_ios-swift--canary.549.5274577056.swift)  
[color_asdk_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_asdk_kotlin--canary.549.5274577056.kt)  
[color_asdk_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_asdk_react-native--canary.549.5274577056.ts)  
[color_sberHealth_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_ios-swift--canary.549.5274577056.swift)  
[color_sberHealth_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_kotlin--canary.549.5274577056.kt)  
[color_sberHealth_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_react-native--canary.549.5274577056.ts)  
[color_sbermarket_business_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_business_ios-swift--canary.549.5274577056.swift)  
[color_sbermarket_business_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_business_kotlin--canary.549.5274577056.kt)  
[color_sbermarket_business_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_business_react-native--canary.549.5274577056.ts)  
[color_sbermarket_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.549.5274577056.swift)  
[color_sbermarket_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.549.5274577056.kt)  
[color_sbermarket_metro_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_metro_ios-swift--canary.549.5274577056.swift)  
[color_sbermarket_metro_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_metro_kotlin--canary.549.5274577056.kt)  
[color_sbermarket_metro_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_metro_react-native--canary.549.5274577056.ts)  
[color_sbermarket_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.549.5274577056.ts)  
[color_sbermarket_selgros_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_selgros_ios-swift--canary.549.5274577056.swift)  
[color_sbermarket_selgros_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_selgros_kotlin--canary.549.5274577056.kt)  
[color_sbermarket_selgros_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_selgros_react-native--canary.549.5274577056.ts)  
[color_sbermarket_wlbusiness_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_wlbusiness_ios-swift--canary.549.5274577056.swift)  
[color_sbermarket_wlbusiness_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_wlbusiness_kotlin--canary.549.5274577056.kt)  
[color_sbermarket_wlbusiness_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_wlbusiness_react-native--canary.549.5274577056.ts)  
[color_sberonline_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberonline_ios-swift--canary.549.5274577056.swift)  
[color_sberonline_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberonline_kotlin--canary.549.5274577056.kt)  
[color_sberonline_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberonline_react-native--canary.549.5274577056.ts)  
[color_sberprime_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.549.5274577056.swift)  
[color_sberprime_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.549.5274577056.kt)  
[color_sberprime_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.549.5274577056.ts)  
[shadow_sbermarket_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/shadow_sbermarket_react-native--canary.549.5274577056.ts)  
[typo_mage_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_ios-swift--canary.549.5274577056.swift)  
[typo_mage_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.549.5274577056.kt)  
[typo_mage_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.549.5274577056.ts)  
[typo_plasma_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_ios-swift--canary.549.5274577056.swift)  
[typo_plasma_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.549.5274577056.kt)  
[typo_plasma_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.549.5274577056.ts)  
[typo_ruler_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_ios-swift--canary.549.5274577056.swift)  
[typo_ruler_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.549.5274577056.kt)  
[typo_ruler_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.549.5274577056.ts)  
[typo_sage_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_ios-swift--canary.549.5274577056.swift)  
[typo_sage_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.549.5274577056.kt)  
[typo_sage_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.549.5274577056.ts)  
[typo_sbermarket_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_ios-swift--canary.549.5274577056.swift)  
[typo_sbermarket_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.549.5274577056.kt)  
[typo_sbermarket_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.549.5274577056.ts)  
[typo_soulmate_ios-swift--canary.549.5274577056.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_ios-swift--canary.549.5274577056.swift)  
[typo_soulmate_kotlin--canary.549.5274577056.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.549.5274577056.kt)  
[typo_soulmate_react-native--canary.549.5274577056.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.549.5274577056.ts)

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-b2c@1.216.1-canary.549.5274577056.0
  npm install @salutejs/plasma-core@1.121.1-canary.549.5274577056.0
  npm install @salutejs/plasma-hope@1.216.1-canary.549.5274577056.0
  npm install @salutejs/plasma-icons@1.150.1-canary.549.5274577056.0
  npm install @salutejs/plasma-temple@1.165.1-canary.549.5274577056.0
  npm install @salutejs/plasma-tokens@1.55.1-canary.549.5274577056.0
  npm install @salutejs/plasma-ui@1.197.1-canary.549.5274577056.0
  npm install @salutejs/plasma-web@1.216.1-canary.549.5274577056.0
  npm install @salutejs/plasma-cy-utils@0.62.1-canary.549.5274577056.0
  npm install @salutejs/plasma-sb-utils@0.119.1-canary.549.5274577056.0
  # or 
  yarn add @salutejs/plasma-b2c@1.216.1-canary.549.5274577056.0
  yarn add @salutejs/plasma-core@1.121.1-canary.549.5274577056.0
  yarn add @salutejs/plasma-hope@1.216.1-canary.549.5274577056.0
  yarn add @salutejs/plasma-icons@1.150.1-canary.549.5274577056.0
  yarn add @salutejs/plasma-temple@1.165.1-canary.549.5274577056.0
  yarn add @salutejs/plasma-tokens@1.55.1-canary.549.5274577056.0
  yarn add @salutejs/plasma-ui@1.197.1-canary.549.5274577056.0
  yarn add @salutejs/plasma-web@1.216.1-canary.549.5274577056.0
  yarn add @salutejs/plasma-cy-utils@0.62.1-canary.549.5274577056.0
  yarn add @salutejs/plasma-sb-utils@0.119.1-canary.549.5274577056.0
  ```
</details>

### CODE REVIEW

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-549/

File: .github/workflows/theme-builder.yml [hunk 1]
```diff
@@ -17,28 +17,7 @@ jobs:
       - name: Prepare repository
```
Comment (Yakutoc) [inline .github/workflows/theme-builder.yml:17]: @neretin-trike @Yeti-or  Парни вопрос, а нам нужен тут этот шаг? На него уходит около - 1m 16s

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-549/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-549/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-549/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-549/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-549/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-549/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-549/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-549/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-549/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-549/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-549/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-549/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-549/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-549/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-549/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-549/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-549/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-549/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-549/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-549/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-549/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-549/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-549/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-549/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-549/temple-storybook/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-549/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-549/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-549/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-549/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-549/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-549/temple-storybook/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-549/

File: .github/workflows/cypress-common.yml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+      - name: Lerna bootstrap
+        run: npx lerna bootstrap
+
+      - name: Run Cypress CT for Plasma ${{ inputs.scope }}
+        if: ${{ success() }}
+        run: npm run cy:${{ inputs.scope }}:run-ct
+
+      - name: Save artifacts
+        if: ${{ failure() }}
```
Comment (Yeti-or) [inline .github/workflows/cypress-common.yml:45]: я бы на мастере вот кстати всегда бы сохранял артифакты хотя бы на месяц, чтобы можно было если что глянуть на отчет сайпруса

**Review by Yeti-or — APPROVED** [review]

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

File: .github/actions/prepare-environment/action.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        key: node-modules-${{ hashFiles('package-lock.json') }}
+        restore-keys: |
+          node-modules-${{ hashFiles('package-lock.json') }}
+          node-modules-
+
+    - name: Cache dependencies
+      id: cache
+      uses: actions/cache@v3
+      with:
+        path: ~/.cache
```
[the remark this answers is not in the document]
Comment (Yakutoc) [reply .github/actions/prepare-environment/action.yml:28]: @Yeti-or 

> это наверно вопрос к тому что ты хотел нам рассказать как работают кэши

В том числе и про это. 

> а что лежит в папке .cache?

```bash
ls ~/.cache

puppeteer Cypress
``` 

там лежат бинарники

File: .github/workflows/cypress-common.yml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+          npm i --no-progress react@17 react-dom@17 @types/react@17.0.40 @types/react-dom@17 --prefix="./packages/plasma-ui"
+          npm i --no-progress react@17 react-dom@17 @types/react@17.0.40 @types/react-dom@17 --prefix="./packages/plasma-temple"
+          npm i --no-progress react@17 react-dom@17 @types/react@17.0.40 @types/react-dom@17 --prefix="./packages/plasma-b2c"
+          npm i --no-progress react@17 react-dom@17 @types/react@17.0.40 @types/react-dom@17 --prefix="./packages/plasma-web"
+
+      - name: Lerna bootstrap
+        run: npx lerna bootstrap
+
+      - name: Run Cypress CT for Plasma ${{ inputs.scope }}
+        if: ${{ success() }}
```
[the remark this answers is not in the document]
Comment (Yakutoc) [reply .github/workflows/cypress-common.yml:41]: @Yeti-or 

Что бы тесты запускались только в том случае когда предыдущий шаг точно завершился с success статусом.

File: .github/workflows/cypress-common.yml (hunk 2 above)
Comment (Yakutoc) [reply .github/workflows/cypress-common.yml:45]: @Yeti-or Объясни пожалуйста для чего?  

> я бы на мастере вот кстати всегда бы сохранял артифакты хотя бы на месяц, 

Давай отдельной задачей?

File: .github/workflows/cypress.yml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+concurrency:
+  # New commit on branch cancels running workflows of the same branch
+  group: ${{ github.workflow }}-${{ github.head_ref }}
+  cancel-in-progress: true
+
+jobs:
+  cypress:
+    strategy:
+      fail-fast: false
```
[the remark this answers is not in the document]
Comment (Yakutoc) [reply .github/workflows/cypress.yml:18]: @Yeti-or 

TL;DR - Что бы остальные job продолжали работу, если какая-нибудь вдруг отвалилась.  

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast

> **fail-fast** applies to the entire matrix. If **fail-fast** is set to true, GitHub will cancel all in-progress and queued jobs in the matrix if any job in the matrix fails. This property defaults to true.

File: .github/workflows/cypress-common.yml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        type: string
+      with-react-17:
+        required: false
+        type: boolean
+        default: false
+
+jobs:
+  run:
+    runs-on: ubuntu-latest
+    continue-on-error: true
```
Comment (Yakutoc) [inline .github/workflows/cypress-common.yml:18]: [Note]

> **continue-on-error** applies to a single job. If **continue-on-error** is true, other jobs in the matrix will continue running even if the job with continue-on-error: true fails.

https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#handling-failures

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-549/

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-549/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-549/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-549/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-549/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-549/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-549/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Review by kayman233 — APPROVED** [review]
